tests/drivers/adc/adc_error_cases/src/main.c

```diff
@@ -0,0 +1,244 @@
+/*
+ * Copyright (c) 2024, Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <zephyr/kernel.h>
+#include <zephyr/device.h>
+#include <zephyr/ztest.h>
+#include <zephyr/drivers/adc.h>
+
+static const struct device *dev_adc = DEVICE_DT_GET(DT_ALIAS(adc));
+#define BUFFER_LEN 8
+uint16_t m_sample_buffer[BUFFER_LEN];
+
+static struct adc_channel_cfg valid_channel_cfg = {
+	.gain = ADC_GAIN_1,
+	.channel_id = 0,
+	.reference = ADC_REF_INTERNAL,
+	.acquisition_time = ADC_ACQ_TIME_DEFAULT,
+	.differential = false,
+	#ifdef CONFIG_ADC_CONFIGURABLE_INPUTS
+	.input_positive = 1,
+	#endif
+};
+
+static struct adc_sequence valid_seq = {
+	.buffer = m_sample_buffer,
+	.buffer_size = BUFFER_LEN * sizeof(uint16_t),
+	.options = NULL,
+	.resolution = 8,
+	.oversampling = 0,
+	.channels = 1,
+};
+
+/**
+ * @brief test adc_read() with invalid oversampling value
+ *
+ * function should return -EINVAL
+ */
+
+ZTEST(adc_error_cases, test_adc_read_invalid_oversampling)
+{
+	int ret;
+
+	adc_channel_setup(dev_adc, &valid_channel_cfg);
+
+	struct adc_sequence invalid_seq = valid_seq;
+	/* Set oversampling to invalid value */
+	invalid_seq.oversampling = 99;
+
+	ret = adc_read(dev_adc, &invalid_seq);
+
+	zassert_true(
+		ret == -EINVAL,
+		"adc_read() should return True,"
+		" got unexpected value of %d",
+		ret
+	);
+}
+
+/**
+ * @brief test adc_read() with invalid resolution value
+ *
+ * function should return -EINVAL
+ */
+
+ZTEST(adc_error_cases, test_adc_read_invalid_resolution)
+{
+	int ret;
+
+	adc_channel_setup(dev_adc, &valid_channel_cfg);
+
+	struct adc_sequence invalid_seq = valid_seq;
+	/* Set resolution to invalid value */
+	invalid_seq.resolution = 99;
```

Comment:
Assuming that valid resolutions are 8-12 bits. I wonder if edge values like 7 or 13 would be better to check. However, this might be incompatible with other vendors (f.e. if their ADC supports higher/lower resolutions).

Reply:
yes, idea was to create generic TCs, as there is no restriction for resolution values in zephyr documentation 7 or 13 might be implemented by other vendors